### PR TITLE
More fine grained routing key settings for messages

### PR DESCRIPTION
**Pull request by TWestling**
Currently, the only way to choose a routing key is to set one which will be used for every message.
This pull request adds an automatic setting of routing keys, where each method that publishes messages gets one based on the method. For details on the automatic routing keys, look at: src/main/webapp/help-automatic-routing-key.html

### Testing done

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

File: src/main/java/com/sonymobile/jenkins/plugins/mq/mqnotifier/MQNotifierConfig.java [hunk 1]
```diff
@@ -291,9 +312,26 @@ public void setVirtualHost(String virtualHost) {
     }
 
     /**
-     * Gets the routing key.
+     * Sets the routing key.
```
Comment (magnusbaeck) [inline src/main/java/com/sonymobile/jenkins/plugins/mq/mqnotifier/MQNotifierConfig.java]:
```suggestion
     * Sets the routing key provider ("AUTO" or "MANUAL").
```

File: src/main/java/com/sonymobile/jenkins/plugins/mq/mqnotifier/MQNotifierConfig.java [hunk 2]
```diff
@@ -291,9 +312,26 @@ public void setVirtualHost(String virtualHost) {
...
+     * Gets the routing key provider.
+     *
+     * @return the routing key provider.
+     */
+
+    public String getRoutingKeyProvider() {
+        return this.routingKeyProvider;
+    }
+    /**
+     * Gets the routing key provider, either "auto" or "manual".
```
Comment (magnusbaeck) [inline src/main/java/com/sonymobile/jenkins/plugins/mq/mqnotifier/MQNotifierConfig.java]:
```suggestion
     * Gets the routing key provider, either "AUTO" or "MANUAL".
```
Should we use constants for these values?

File: src/main/java/com/sonymobile/jenkins/plugins/mq/mqnotifier/pipeline/MQMessageStep.java [hunk 3]
```diff
@@ -107,7 +107,7 @@ protected Void run() throws Exception {
             if (config.getEnableVerboseLogging()) {
                 listener.getLogger().println("Posting JSON message to RabbitMQ:\n" + json.toString(2));
             }
-            MQConnection.getInstance().publish(json);
+            MQConnection.getInstance().publish(json, "publishMQMessage");
```
Comment (magnusbaeck) [inline src/main/java/com/sonymobile/jenkins/plugins/mq/mqnotifier/pipeline/MQMessageStep.java]: A potential follow-up could be to add the routing key as an optional argument to the pipeline step.

File: src/main/webapp/help-manual-routing-key.html [hunk 4]
```diff
@@ -0,0 +1 @@
+<p>With this setting, every message sent will have the routing key specified in the "Routing Key" field.</p>
```
Comment (magnusbaeck) [inline src/main/webapp/help-manual-routing-key.html]: Missing newline at EOF.

**Review by magnusbaeck — CHANGES_REQUESTED** [review]
Code LGTM, just a few comment nits that we might as well fix.

Side note: We're not covering this new functionality in any tests.

**Comment by TWestling** [comment]
Added tests + fixed nits. Also added the optional routing key for the message step since it felt like very low hanging fruit.

File: src/test/java/com/sonymobile/jenkins/plugins/mq/mqnotifier/PluginTest.java [hunk 5]
```diff
@@ -41,9 +40,11 @@
 
 import static org.hamcrest.CoreMatchers.containsString;
 import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
-import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertNotEquals;
+
```
Comment (magnusbaeck) [inline src/test/java/com/sonymobile/jenkins/plugins/mq/mqnotifier/PluginTest.java:47]: Extra blank line.

**Review by magnusbaeck — APPROVED** [review]

**Pull request merged**